Major Update: Refactored components, new context, and enhanced editor text

juliopavila opened the pull request:
## Description

This PR represents a major update to the Tabula project, introducing significant changes and improvements. We've refactored components for performance, split logic into custom hooks for better organization, introduced a new context for articles, and revamped our editor text component with rich text and inline text capabilities using Draft.js.

---

Below are the key changes:

### Editor Text Component with Rich Text and Inline Text
We've created a new Editor Text component that supports rich and inline text. This was done using Draft.js as a block editor. This will enhance the writing experience and allow more complex formatting within the text editor.

### Performance Refactoring of the Deterministic Avatar Component
The Deterministic Avatar component has been refactored with an emphasis on performance improvement. This should lead to smoother interactions and improved load times within the application.

### Custom Hooks
We've reorganized some of the logic into custom hooks. This design approach makes our components leaner, more manageable, and easier to understand.

### Article Context Creation
The creation of the Article Context introduces a structured way of managing and passing around article data across different components. This greatly improves our state management for articles.

### Removal of Old Article Components
As a part of our codebase cleanup and modernization, we've removed the old versions of article creation components. This is part of our effort to streamline the codebase and improve maintainability.

The PR contributes to enhancing the core functionalities and performance of the application while laying a more solid foundation for future features and improvements.

Please review and provide any feedback or questions you might have.

---

### Dependencies
This PR introduces some new dependencies, including Draft.js for the editor text component, and has updated versions of several others. The `package.json` has been updated accordingly.

---

### Testing
For this PR, extensive manual end-to-end (E2E) testing was carried out to ensure the correct operation of the newly introduced features and changes. This was in addition to the existing test suite. Please note that updates to automated tests have not been included in this PR, and all testing was conducted manually to ensure thorough and comprehensive coverage.

---

### Changelog

- Created a new Editor Text Component with Draft.js
- Refactored the Deterministic Avatar Component for performance
- Split logic into custom hooks
- Created a new Article Context
- Removed old versions of creating article components
- Updated dependencies
- Manual e2e testing
- Refactor ipfs handler
- Implemented a new function, `uploadToInfura`, to upload a file or string content to IPFS via the Infura API. This function also handles the conversion of strings to blobs and the authorization headers for Infura. The `pin=false` query parameter is appended to the Infura API request, which prevents the file from being pinned by default.
- The `uploadContent` function now has a try-catch block that attempts to upload the content to IPFS using the IPFS HTTP client. If this fails, it falls back to the `uploadToInfura` function. This helps handle scenarios where the user is using a public IPFS gateway, which might not support generating a CID.
- In order to use the Infura API, you will need to add your Infura API Key and Secret to your environment variables. Please add the following environment variables to your `.env` file:

```plaintext
REACT_APP_INFURA_API_KEY=your_infura_api_key
REACT_APP_INFURA_API_KEY_SECRET=your_infura_api_key_secret
```

cedricwaxwing committed:
Adds styles for Sidebar and CreateArticle Page/Header

juliopavila committed:
Create Rich Text Component

cedricwaxwing committed:
Header/Sidebar layout adjusments

juliopavila committed:
Create editor block component

juliopavila committed:
Refactor editable block component

juliopavila committed:
update yarn lock

cedricwaxwing committed:
Updates contenteditable styles

cedricwaxwing committed:
Merge branch 'main' into create-article-updates

cedricwaxwing committed:
Merge branch 'create-article-updates' into editor-block-component

juliopavila committed:
Add image tag

juliopavila committed:
Merge branch 'editor-block-component' of ssh://github.com/gnosis/tabula into editor-block-component

juliopavila committed:
Solved merge conflicts

cedricwaxwing committed:
Updates image block style

cedricwaxwing committed:
Only show setting icon when sidebar is not active.

juliopavila committed:
feat: add divider component, and close menu clicking outside

cedricwaxwing committed:
Updates user dropdown

cedricwaxwing committed:
Update ArticleSidebar scrolling/overflows

cedricwaxwing committed:
Updates editor typography/element styles

cedricwaxwing committed:
refine heading typography

cedricwaxwing committed:
WIP - adds initial Inline Style UI

juliopavila committed:
feat: Create/Edit Article improvements

juliopavila committed:
Merge branch 'editor-block-component' of ssh://github.com/gnosis/tabula into editor-block-component

juliopavila committed:
Solved warnings

cedricwaxwing committed:
Update ArticleHeader.tsx

cedricwaxwing committed:
updates turndown location

cedricwaxwing committed:
Delete yarn.lock

cedricwaxwing committed:
updates yarn.lock

cedricwaxwing committed:
Update yarn.lock

cedricwaxwing committed:
Temporarily removes author info from article

cedricwaxwing committed:
WIP - Adds edit button to publication page

cedricwaxwing commented on packages/app/src/components/views/publication/ArticleView.tsx:
```diff
@@ -77,7 +77,7 @@ export const ArticleView: React.FC<ArticleViewProps> = ({ updateChainId }) => {
 
   return (
-    <PublicationPage showCreatePost={false} publication={article?.publication}>
+    <PublicationPage showCreatePost={false} showEditButton={true} publication={article?.publication}>
```
@juliopavila I hardcoded this as true for now, but it will need to be dynamic based on whether the user has edit permissions

cedricwaxwing committed:
Adds overflowWrap to article view

juliopavila committed:
feat: improve create/edit/read article

juliopavila committed:
Merge branch 'editor-block-component' of ssh://github.com/gnosis/tabula into editor-block-component

cedricwaxwing committed:
Merge branch 'editor-block-component' of https://github.com/gnosis/tabula into editor-block-component

juliopavila committed:
feat: Improve how we transform html to article blocks

juliopavila committed:
feat: add new row button and implement quote/code tags

cedricwaxwing committed:
minor style updates

cedricwaxwing committed:
Resize upload button

juliopavila committed:
feat: add edit article from article view page

juliopavila committed:
Merge branch 'main' into editor-block-component

juliopavila committed:
solve build issue

cedricwaxwing commented:
Remaining bugs:

When we add an image in the rich text editor, it seems to make it impossible to add any other text. Also, it should be possible to select it with your cursor and click delete, like we can do for the text. (This is way lower priority though)
- [ ] When the "publish" is complete, it should take you to the article page automatically.
- [ ] I would like to add the captions for the thumbnail and rich text, but if this is complex, let's skip for now.
- [ ] If someone just posts an image, or images, it should still allow them to publish. Currently it stops and it says "Article Content is Required"
-  [ ] If you go back to publications after getting an error and click "New Post" again, it should not show the old error on the page.
- [ ] Clicking preview did not show the thumbnail I had added in the sidebar. It also removed my content from Article Content.
- [ ] For the loading state on the article page, we should add some text below the loader that says something like: "decrypting data from IPFS...please wait a moment"
- [ ] I tried uploading a 4MB thumbnail, and I think it breaks. We should maybe validate a max file size if this will break.

juliopavila committed:
feat: improve the UX when they wanna create a new article

juliopavila committed:
feat: edit article from Article view screen

juliopavila committed:
feat: improve of how to handle the update article flow

cedricwaxwing committed:
Removes sans styling from preview title

juliopavila committed:
fix: Solved build issue

juliopavila committed:
[Feat]: Improve performance entire app, refactor deterministic avatar, create article context

asgeir-s reviewed (commented):
This looks amazing!

Some feedback:

- [ ]  The edit box pops up off-screen and can not be seen (if used on a not-very-wide screen):

<img width="300" alt="Screenshot 2023-05-09 at 13 38 43" src="https://github.com/gnosis/tabula/assets/2156509/5db31d50-d0f4-4956-a221-5fafea875e31"> Here the box is open, but can not be seen it's off-screen.
- [ ]   I don't know if it should be possible, but it's not possible to write `/h1` to convert the current component to an `h1` component.
- [ ]   If a text component (paragraph, code snippet, quote) is converted into an image or a divider component, it's no way to undo it; the text is lost `cmd-z` does not work then.
- [ ]   When an image component with an image is tuned into another component type, it's not possible to `cmd-z` back to the image.
- [ ]   When undoing, an empty component is kept for each line.
- [ ]   The thumbnail image is displayed at the top of the post and it's a thumbnail for the post. Perhaps we should rename it to "article image"?
<img width="858" alt="Screenshot 2023-05-09 at 13 54 19" src="https://github.com/gnosis/tabula/assets/2156509/55958dd4-9251-4281-8b22-5de120251fcd">

- [x]   When clicking the publication in the new edit page, the publicationId is used instead of the slug, this makes the URL transform from `http://localhost:3000/#/forplay.eth/...` into `http://localhost:3000/#/5-P-0x45207e251e415ee08d5b374f2f344b5ffa698d4b11ec7ac76f83a1c620a47a07-356/...`: 
<img width="870" alt="Screenshot 2023-05-09 at 14 02 37" src="https://github.com/gnosis/tabula/assets/2156509/fa2de670-3072-46ca-b7cb-8b5ac3e77d92">

- [x]   No feedback when trying to publish without content or title. When clicking Publish, nothing happens.

juliopavila committed:
bug: solve empty state process

manboy-eth commented on packages/app/src/components/layout/ArticleHeader.tsx:
```diff
@@ -0,0 +1,347 @@
+          sx={{ cursor: "pointer", transition: "opacity 0.25s ease-in-out", "&:hover": { opacity: 0.6 } }}
+          onClick={handleNavigation}
+        >
+          <Avatar width={31} height={31} publicationSlug={publication?.id} />
```
```suggestion
          <Avatar width={31} height={31} publicationSlug={publicationSlug ?? publication?.id} />
```

manboy-eth commented on packages/app/src/components/layout/CreateArticlePage.tsx:
```diff
@@ -0,0 +1,119 @@
+          <meta property="og:description" content={publication?.description} key="1" />,
+          <meta name="description" content={publication?.description} key="2" />,
+        ]}
+        <meta property="og:url" content={`https://tabula.gg/#/${publication?.id}`} />
```
Should be publication slug

manboy-eth commented on packages/app/src/components/layout/CreateArticlePage.tsx:
```diff
@@ -0,0 +1,119 @@
+const PublicationPage: React.FC<Props> = ({ children, publication, type }) => {
+  const location = useLocation()
+  const [showSidebar, setShowSidebar] = useState<boolean>(true)
+  const { imageSrc } = usePublication(publication?.id || "")
```
Should use publication slug here

manboy-eth commented on packages/app/src/components/views/publication/ArticleView.tsx:
```diff
@@ -1,32 +1,45 @@
+  const { article, saveArticle, markdownArticle, setMarkdownArticle, loading, getIpfsData } = useArticleContext()
+
   const { data, executeQuery, imageSrc } = useArticle(articleId || "")
   const publication = usePublication(article?.publication?.id || "")
```
Should use publication slug here

asgeir-s reviewed (commented):
We can probably delete the old unused article creation views (`CreateArticleView2 `) now, right?

manboy-eth commented on packages/app/src/components/views/publication/PublicationView.tsx:
```diff
@@ -103,25 +104,14 @@ export const PublicationView: React.FC<PublicationViewProps> = ({ updateChainId
-                        {publication.id}
-                      </Typography>
-                    </Stack>
+                      {publication.id}
```
```suggestion
                      {publicationSlug}
```

manboy-eth commented on packages/app/src/constants/chain.ts:
```diff
@@ -22,6 +29,25 @@ export enum SupportedChain {
+  case SupportedChainId.POLYGON: 
+    return PolygonIcon
+  case SupportedChainId.ARBITRUM: 
+    return ArbitrumtIcon
```
```suggestion
    return ArbitrumIcon
```

manboy-eth commented on packages/app/src/hooks/useIpfs.ts:
```diff
@@ -4,7 +4,7 @@ import axios from "axios"
 import { getClient } from "../services/ipfs"
 
-const IPFS_GATEWAY = process.env.REACT_APP_IPFS_GATEWAY
+export const IPFS_GATEWAY = process.env.REACT_APP_IPFS_GATEWAY
```
```suggestion
const IPFS_GATEWAY = process.env.REACT_APP_IPFS_GATEWAY
```
This should not be exported and never used directly. 
It's a fallback that will only be used if no IPFS node is provided by the user. 

We should strive to use the one provided by the user via the useIpfs hook. If, however, we require a URL for, for instance, getting an image, we should use the exported wrapper functions instead, like `getText` and `getImgSrc`, or create a new one if required.

FYI: `getImgSrc` is right now using the `IPFS_GATEWAY`, but this is a provisional workaround that will be fixed, so we should still use the `getImgSrc`.

manboy-eth commented on packages/app/src/hooks/useMarkdown.tsx:
```diff
@@ -0,0 +1,73 @@
+          image = URL.createObjectURL(block.imageFile)
+        }
+
+        if (block.imageUrl && IPFS_GATEWAY) {
+          if (addUrlPrefix && !block.imageUrl.includes(IPFS_GATEWAY)) {
+            image = await getImgSrc(block.imageUrl)
+          } else {
+            image = block.imageUrl
+          }
+        }
+        if (block.imageUrl && isDraft && IPFS_GATEWAY && type === "read" && block.imageUrl.includes(IPFS_GATEWAY)) {
+          image = block.imageUrl
+        }
+        if (block.imageUrl && IPFS_GATEWAY && block.imageUrl.includes(IPFS_GATEWAY) && type === "publish") {
+          image = removePrefix(block.imageUrl, `${IPFS_GATEWAY}/`)
+        }
+      }
```
Update to not use IPFS_GATEWAY directly (see last comment)

juliopavila committed:
feat: apply PR comments

asgeir-s reviewed (changes requested):
Nice work! Most things have been fixed.

However, some stuff here still requires handling: https://github.com/gnosis/tabula/pull/217#pullrequestreview-1418511427 (I added checkmarks to make it easy to see what's missing).

cedricwaxwing committed:
Update vertical alignment of menu on smaller sizes and removes unfocused placeholder lines

juliopavila committed:
feat: Create new editor component with draftJS

cedricwaxwing committed:
Initial inline style menu positioning updates

juliopavila committed:
feat: implement divider component

cedricwaxwing committed:
Resolves OL and UL styling issues

juliopavila committed:
feat: Add link changes

juliopavila committed:
bug: solve link  issue

juliopavila committed:
chore: remove old code

juliopavila committed:
chore: remove warning

juliopavila committed:
feat: create image component

juliopavila committed:
bug: Duplicate transaction creation

juliopavila committed:
bug: solved update article content without images

juliopavila committed:
bug: solution article hash update

juliopavila committed:
Merge branch 'editor-block-component' into inline-text-component

juliopavila committed:
Merge pull request #227 from gnosis/inline-text-component

Refactor Editor Rich text

juliopavila committed:
chore: remove unnecessary dependencies

juliopavila committed:
chore: clean general package json

asgeir-s reviewed (commented):
Nice, I can feel the performance improvements 🔥 

Some feedback:

- [x] Links do not become clickable links in the preview or when published.

- [ ] When clicking edit on a published article, I get an error sample article: http://localhost:3000/#/forplay.eth/5-A-0x1e22ca9088459ce943a89f9c9d01b1cad0999af64cfed328749e7f54c6776907-78. 
<img width="1154" alt="Screenshot 2023-06-02 at 16 15 57" src="https://github.com/gnosis/tabula/assets/2156509/a46f1f82-0244-44b4-8b41-5b754c2f56f2">

- [x] It looks like the edit article button in the article list is always disabled: 

<img width="961" alt="Screenshot 2023-06-02 at 16 17 00" src="https://github.com/gnosis/tabula/assets/2156509/9c74bfc4-727a-4260-a1ff-a7f34f501e87">

- [x] The editor popup menu still does not show up on smaller screens; it is outside of view. We at least need a quick fix for this.

**New issues added 9. Jun**:
- [x]  No articles can be opened from the article list.

- [x] After entering the link, it's not possible to click enter, and if I click outside of the box, nothing happens (the link just disappears).

juliopavila committed:
feat: Deselect the text after inline selection

juliopavila committed:
feat: solved cursor should be focused in the contenteditable input and ipfs issue

juliopavila committed:
bug: solve link url persistence after inline component close

juliopavila committed:
bug: solved code-block component

juliopavila committed:
revert Ipfs change

juliopavila committed:
bug: solved strikethrough issue

juliopavila committed:
bug: solved code inline style in the editor component

juliopavila committed:
bug: show article content error when is empty

juliopavila committed:
feat: refactor ipfsHttpClient to handle personal node, public node and local node

juliopavila committed:
feat: change IPFS order

asgeir-s committed:
Only connect to a public IPFS node if the in-browser-ipfs-node-setup fails

asgeir-s committed:
Pluralize the articles section

juliopavila committed:
feat: change target for link component

juliopavila committed:
feat: show placeholder for every empty and focus block

juliopavila committed:
feat: improve underline for markdown

juliopavila committed:
bug: show article content if the article content coming without hash

juliopavila committed:
bug: solved double transaction

juliopavila committed:
feat: improve link flow

cedricwaxwing committed:
Resolves left offset issue

cedricwaxwing committed:
Adds top offset logic back

cedricwaxwing committed:
Handle positioning for showUrlnput and left offset

juliopavila committed:
chore: add useEffect dependecies

juliopavila committed:
chore: comment browser ipfs node because ipfs-js is deprecated

asgeir-s reviewed (changes requested):
I see that most of the issues a fixed. However, I found two issues that needs to be fixed:

### Issue 1
When clicking the "Edit" button (both inside the article page and in the article list), I get this error, and the application freezes (I have to reload the page to make it interactive again): 
<img width="1079" alt="Screenshot 2023-06-28 at 13 25 14" src="https://github.com/gnosis/tabula/assets/2156509/8a5ec8bb-ff28-43b9-9362-948be8695b2b">
This is especially bad when writing an article since you will lose all your writing.

**Reproduce** 
1. Open the "article list" page or an article page.
2. Click the "Edit" button.

### Issue 2
The app freezes after some time and returns this error (I have to refresh the page to make it interactive again):
<img width="1080" alt="Screenshot 2023-06-28 at 13 28 11" src="https://github.com/gnosis/tabula/assets/2156509/2194b4ad-b10a-4765-b7ae-71d40925f507">
This is especially bad when writing an article since you will lose all your writing.

**Reproduce** 
1. Open the "article list" page, an article page, or the "create new article" page in the app.
2. Wait for a couple of minutes.
3. The error shows up in the console up, and the UI freezes.

juliopavila committed:
bug: Add error handling to fetchArticleContent function

juliopavila committed:
bug: handle invalid hash content

asgeir-s reviewed (changes requested):
Found one issue.

Publishing fails from the article's preview page. And when going back to edit, the article content is empty.

#### Reproduce:
1. Click edit on an article (for example, the `my article is cool` article in this publication: http://localhost:4000/#/5-P-0x45207e251e415ee08d5b374f2f344b5ffa698d4b11ec7ac76f83a1c620a47a07-356).
2. Click the Preview button.
3. In the Preview page, click the Publish button.
4. Nothing happens.
5. If you click the Edit button in the preview page, the content will be empty.

juliopavila committed:
feat: If the upload fails, it attempts to upload the file using the Infura API

juliopavila commented:
> Found one issue.
> 
> Publishing fails from the article's preview page. And when going back to edit, the article content is empty.
> 
> #### Reproduce:
> 1. Click edit on an article (for example, the `my article is cool` article in this publication: http://localhost:4000/#/5-P-0x45207e251e415ee08d5b374f2f344b5ffa698d4b11ec7ac76f83a1c620a47a07-356).
> 2. Click the Preview button.
> 3. In the Preview page, click the Publish button.
> 4. Nothing happens.
> 5. If you click the Edit button in the preview page, the content will be empty.

@manboy-eth This is a workaround to the issue of generating a CID for IPFS content. We're now leveraging the Infura API when necessary, especially in cases where a public IPFS gateway fails to generate a CID.

I've introduced a new function called `uploadToInfura` that uploads a file or string content to IPFS via the Infura API. If the file is a string, the function converts it to a blob before making the API request. We've also set `pin=false` in the API request to avoid pinning the file by default.

For this function to work properly, we need to supply it with the Infura API Key and Secret. I've added the placeholders for these keys to our environment variables. Please update your `.env` file with your Infura API Key and Secret:

```plaintext
REACT_APP_INFURA_API_KEY=your_infura_api_key
REACT_APP_INFURA_API_KEY_SECRET=your_infura_api_key_secret
```

This new function is used in our `uploadContent` function as a fallback when the IPFS HTTP client fails to upload the content. This provides a reliable way to ensure our content gets uploaded to IPFS.

asgeir-s reviewed (approved):
Nice work @juliopavila 🚀
